sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java

```diff
@@ -0,0 +1,390 @@
+        return (isLastBatch) -> {
+            CompletableFuture<BitSet> executeBulkFuture = new CompletableFuture<>();
+
+            if ((isLastBatch && pendingRequestsByIndex.isEmpty() == false)
+                || pendingRequestsByIndex.size() > createIndicesBulkSize) {
+
+                // We create new indices in an async fashion and execute the requests after they are created.
+                // This means the iterator consumer thread can carry on accumulating rows for the next bulk.
+                // We copy and clear the current bulk global requests and pendingRequests structures in order to allow the
+                // iterator consumer thread to create the next bulk whilst we create indices and execute this bulk.
+                Map<String, List<PendingRequest<TItem>>> pendingRequestsForCurrentBulk =
+                    new HashMap<>(pendingRequestsByIndex);
+                pendingRequestsByIndex.clear();
+
+                Map<ShardLocation, TReq> bulkRequests = new HashMap<>(requestsByShard);
+                requestsByShard.clear();
+
+                createPendingIndices(pendingRequestsForCurrentBulk)
+                    .thenAccept(resp -> {
+                        bulkRequests.putAll(getFromPendingToRequestMap(pendingRequestsForCurrentBulk));
+                        sendRequestsForBulk(executeBulkFuture, bulkRequests);
+                    });
+            } else {
+                sendRequestsForBulk(executeBulkFuture, requestsByShard);
+            }
+
+            return executeBulkFuture;
+        };
+    }
+
+    @Override
+    public CompletableFuture<? extends Iterable<Row>> apply(BatchIterator batchIterator) {
+        CompletableFuture executionFuture = new CompletableFuture();
+        new BatchIteratorBackpressureExecutor<>(batchIterator, scheduler,
+            rowConsumer, execute, backpressureTrigger, pendingItemsCount, bulkSize, BACKOFF_POLICY, executionFuture).
+            consumeIteratorAndExecute();
+
+        return executionFuture.
+            thenApply(ignored -> Collections.singletonList(new Row1((long) responses.cardinality())));
+    }
+
+    @Nullable
+    private ShardLocation getShardLocation(String indexName, String id, @Nullable String routing) {
+        try {
+            ShardIterator shardIterator = clusterService.operationRouting().indexShards(
+                clusterService.state(),
+                indexName,
+                id,
+                routing
+            );
+
+            String nodeId;
+            ShardRouting shardRouting = shardIterator.nextOrNull();
+            if (shardRouting == null || shardRouting.active() == false) {
+                nodeId = shardRouting.relocatingNodeId();
+            } else {
+                nodeId = shardRouting.currentNodeId();
+            }
+
+            if (nodeId == null) {
+                LOGGER.debug("Unable to get the node id for index {} and shard {}", indexName, id);
```

Comment:
consider wrapping into `LOGGER.isDebugEnabled` to avoid varargs array allocation